### PR TITLE
AsyncCallback Iterator To Stream Internal Events

### PR DESCRIPTION
**Pull request by jordanparker6**
# Description

Llama Index should have an out-of-the-box callback that allows the user to stream all internal events through an async iterator. 

The current streaming misses all the internal events like tool usage and agent actions.

This async iterator should also stream token events as well.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

This is still in draft and needs input and fixes.

### CODE REVIEW

File: llama_index/callbacks/async_iter.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+from llama_index.callbacks.utils import superjson_dumps
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncIteratorCallbackHandler(BaseCallbackHandler):
+    """Callback handler that returns an async iterator."""
+
+    queue: asyncio.Queue[str]
+    done: asyncio.Event
```
[the remark this answers is not in the document]
Comment (jordanparker6) [reply llama_index/callbacks/async_iter.py:16]: Fixed this by setting done at the end of the trace. Still working through issues with this just trying to get something up.

[the remark this answers is not in the document]
Comment (jordanparker6) [reply llama_index/callbacks/async_iter.py:16]: Could also leave it up to the user to set...

**Comment by logan-markewich** [comment]
@jordanparker6 are you still picking away at this? If not one of us can probably figure this out sometime in the coming week. It feels like a useful feature :)

**Comment by jordanparker6** [comment]
Hey,

Sorry, I have add to move to high priority tickets. Was going to get back to it in time but happy for you guys to take it on.

Jordan

File: llama_index/callbacks/async_iter.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        event_id: str = "",
+        parent_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        """Run when an event starts and return id of event."""
+        event = {
+            "event": event_type + ".start",
+            "event_id": event_id,
+            "parent_id": parent_id,
+        }
```
Comment (tbeamish-benchsci) [inline llama_index/callbacks/async_iter.py:45]: Why not put the payload in here as well? That way whatever is reading from the queue will get all the event details.

**Comment by logan-markewich** [comment]
Going to close this one out for now, its quite outdated at this point, especially given instrumentation has been introduced to replace the callbacks system

If its still an issue, please raise an updated PR

Sorry for letting this one sit for so long!

**Pull request closed without merging**